examples/common/convert_c4.py

```diff
@@ -4,105 +4,353 @@
 """C4 streaming dataset conversion scripts."""
 import os
 import platform
-from argparse import ArgumentParser, Namespace
-from typing import Dict, Iterable
+import warnings
+from argparse import Action, ArgumentError, ArgumentParser, Namespace
+from enum import Enum
+from textwrap import dedent
+from typing import Dict, Iterable, Optional
 
 import datasets as hf_datasets
+import numpy as np
 from streaming import MDSWriter
 from torch.utils.data import DataLoader, IterableDataset, get_worker_info
 from tqdm import tqdm
+from transformers import AutoTokenizer, PreTrainedTokenizerBase
+
+
+class ConcatMode(Enum):
+    NO_CONCAT = 'NO_CONCAT'
+    CONCAT_TEXT = 'CONCAT_TEXT'
+    CONCAT_TOKENS = 'CONCAT_TOKENS'
+
+
+class EmptyOrNonexistentDirectory(Action):
+
+    def __call__(self, parser, namespace, values, option_string=None):
+        prospective_dir = values
+        if os.path.isdir(prospective_dir) and len(
+                os.listdir(prospective_dir)) > 0:
+            raise ArgumentError(self,
+                                f'--out_root={prospective_dir} must be empty')
+        else:
+            setattr(namespace, 'out_root', prospective_dir)
```

Comment:
I feel like most ppl just parse the args then check and raise a ValueError, or something, but this is fancy AF ❤️

Reply:
Also, I think this is a bit too strict, like if you write the `val` split you should still be allowed to later write the `train` split?

Reply:
had to remove this beauty to let it be less strict :P

Reply:
Abhi is right, but I have to remove the fancy `Action` subclass, since it can't validate against other args (we need to know splits) since it hasn't parsed that yet :sob:

Reply:
@dakinggg  beat me to it